### PR TITLE
feat: HEXA-1120 template pipelines and template versions models + endpoint to create template

### PR DESCRIPTION
**Pull request by YolanFery**
Introduce models for PipelineTemplate and PipelineTemplate versions
Create a first endpoint to create a PipelineTemplate from a Pipeline

## Changes

- New  `PipelineTemplate` and `PipelineTemplateVersion`
- First version of the GraphQL Schema
- Add `get_or_create_template` to the Pipeline model
- Unit tests for the new models and method
- Integ test for the GraphQL Schema
- Register `PipelineTemplateVersion` and `PipelineTemplate` in the admin panel

## How/what to test

This is laying foundations for the Template Pipeline feature and should have no impact on the rest

### CODE REVIEW

File: hexa/template_pipelines/models.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        ]
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    created_at = models.DateTimeField(auto_now_add=True)
+    version_number = models.PositiveIntegerField(editable=False)
+    template = models.ForeignKey(
+        Template, on_delete=models.CASCADE, related_name="versions"
+    )
+    source_pipeline_version = models.OneToOneField(
+        PipelineVersion, on_delete=models.RESTRICT, related_name="template_version"
```
Comment (YolanFery) [inline hexa/template_pipelines/models.py:94]: Restricting deletion for now, we might think about a copying mechansim on delete

File: hexa/template_pipelines/graphql/schema.graphql [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+"""
+Represents the input for creating a new template version.
+"""
+input CreateTemplateVersionInput {
+  name: String  # The name of the template.
+  code: String  # The code of the template.
+  description: String  # The description of the template.
+  config: String  # The configuration of the template.
+  workspace_slug: String!  # The slug of the workspace.
+  pipeline_id: ID!  # The ID of the pipeline.
```
[the remark this answers is not in the document]
Comment (qgerome) [reply hexa/template_pipelines/graphql/schema.graphql:17]: Graphql strings are camelCased

**Review by qgerome — COMMENTED** [review]
`Template` is IMO too broad and may lead to confusion later when we want to templates  for another object.

I'd advice to prefix it by `Pipeline` to avoid clashes in the future. WDYT?

File: hexa/pipelines/models.py [hunk 3]
```diff
@@ -457,6 +458,20 @@ def merge_pipeline_config(
         merged_config = {**cleaned_pipeline_version_config, **cleaned_provided_config}
         return merged_config
 
+    def get_or_create_template(self, name, code, description, config):
+        if not hasattr(self, "template") or not self.template:
+            Template = apps.get_model("template_pipelines", "Template")
```
[the remark this answers is not in the document]
Comment (YolanFery) [reply hexa/pipelines/models.py]: Yes, we can take this approach, thanks.

Changed in next revisions

File: hexa/pipeline_templates/schema/mutations.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    workspace = get_workspace(request.user, input.get("workspace_slug"))
+    if not workspace:
+        return {"success": False, "errors": ["WORKSPACE_NOT_FOUND"]}
+
+    if not request.user.has_perm(
+        "pipeline_templates.create_pipeline_template_version", workspace
+    ):
+        raise PermissionDenied()
+
+    source_pipeline = get_source_pipeline(request.user, input.get("pipeline_id"))
```
[the remark this answers is not in the document]
Comment (YolanFery) [reply hexa/pipeline_templates/schema/mutations.py]: Thanks

File: hexa/pipeline_templates/schema/mutations.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    input = kwargs["input"]
+
+    workspace = get_workspace(request.user, input.get("workspace_slug"))
+    if not workspace:
+        return {"success": False, "errors": ["WORKSPACE_NOT_FOUND"]}
+
+    if not request.user.has_perm(
+        "pipeline_templates.create_pipeline_template_version", workspace
+    ):
+        raise PermissionDenied()
```
[the remark this answers is not in the document]
Comment (YolanFery) [reply hexa/pipeline_templates/schema/mutations.py:45]: Thanks

File: hexa/pipeline_templates/models.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                name="unique_template_name",
+                condition=Q(deleted_at__isnull=True),
+            ),
+        ]
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    name = models.CharField(max_length=200, null=True, blank=True)
```
[the remark this answers is not in the document]
Comment (YolanFery) [reply hexa/pipeline_templates/models.py]: Thanks, it's a mistake, corrected in next commit

**Review by nazarfil — APPROVED** [review]

**Pull request merged**